sdcm/provision/aws/capacity_reservation.py

```diff
@@ -0,0 +1,241 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright (c) 2024 ScyllaDB
+import logging
+import time
+from datetime import datetime, timedelta
+from typing import List, Dict, Tuple
+
+from botocore.exceptions import ClientError
+import boto3
+
+LOGGER = logging.getLogger(__name__)
+
+
+# Shlould runner be created in the same az always? - in that case CR needs to be created before sct-runner $0.01/GB data transfer price
+
+
+class CapacityReservationError(Exception):
+    pass
+
+
+class SCTCapacityReservation:
+    """Class for managing capacity reservations for AWS instances.
+
+    Serves namespacing for capacity reservations and provides methods for creating and cancelling reservations."""
+    reservations: Dict[str, Dict[str, str]] = {}
+
+    @staticmethod
+    def _get_cr_request_based_on_sct_config(params) -> Tuple[dict[str, int], int]:
+        instance_counts = {}
+        nemesis_node_count = params.get("nemesis_add_node_cnt") or 0
+        cluster_max_size = (params.get("cluster_target_size") or params.get("n_db_nodes")) + nemesis_node_count
+        instance_counts[params.get("instance_type_db")] = instance_counts.get(
+            params.get("instance_type_db"), 0) + cluster_max_size
+        instance_counts[params.get("instance_type_loader")] = instance_counts.get(params.get("instance_type_loader"),
+                                                                                  0) + params.get("n_loaders")
+        # don't reserve capacity for monitor - as usually it's not a problem to spin it
+        duration = params.get("test_duration") + 60  # 60 to have margin for test setup
+        instance_counts = {k: v for k, v in instance_counts.items() if v > 0}  # remove 0 values
+        return instance_counts, duration
+
+    @classmethod
+    def get_cr_from_aws(cls, params) -> None:
+        """Retrieves capacity reservations for given test_id from AWS."""
+        if not cls.is_capacity_reservation_enabled(params):
+            LOGGER.info("Capacity reservation is not enabled. Skipping reservation.")
+            return
+        test_id = params.get("reuse_cluster") or params.get("test_id")
```

Comment:
Are you sure about the `reuse_cluster` value?
Have you really verified this case?

Reply:
`reuse_cluster` is taken into account at later stages of a test - not in `sct_config` where `get_cr_from_aws` is used.
This could lead to miss used capacity in case of reusing cluster.

I don't expect to use capacity reservations and reusing cluster case anyway, this is just a form of protection

Reply:
reuse_cluster holds test_id when is used

there's no point to reuse cluster and set capacity reservation to true - if cluster was created (so it can be reused), there's no point in setting it.

Reply:
No one is going to keep in mind to change some config options.
People will just reuse config files.
So, argument `no point` will be violated in practice.

Reply:
Reuse cluster in this case should use the reservation that should exist , and continue to the step

We need just to validate that exactly what's gonna happen

And that reuse works, we don't test the reuse cases regularly, we should take the opportunity 😉

An idea, add reuse case to on or all of the provision tests

Reply:
this looks like separate feature to add it to SCT test pipeline. Also would need to add capacity_reservation setting to be able to test it with it.
I'll do tests manually.

Reply:
I verified it manually - works.
Need to remember that CR is for specified time duration (max test duration + 1h), so it becomes unusable after that time (but test will work, as default behavior is to skip CR params if CR's are not found for given test). It will cause problems if CR caused fallback to another AZ - in that case need to set in test config AZ manually when reusing cluster.